block.py

```diff
@@ -4,100 +4,103 @@
-        ## In inspect-ts it needed to be converted
-        block_receipts_raw = base_provider.make_request("eth_getBlockReceipts", [block_number])
-
-        ## Trace the whole block, return those calls
-        block_calls = w3.parity.trace_block(block_number)
-        
-        ## Get the logs
-        block_hash = (block_data.hash).hex()
-        block_logs = w3.eth.get_logs({'blockHash': block_hash})
-
-        ## Get gas used by individual txs and store them too
-        txs_gas_data = {}
-        for transaction in block_data['transactions']:
-            tx_hash = (transaction.hash).hex()
-            tx_data = w3.eth.get_transaction(tx_hash)
-            tx_receipt = w3.eth.wait_for_transaction_receipt(tx_hash)
-            txs_gas_data[tx_hash] = {
-                'gasUsed': tx_receipt['gasUsed'], # fix: why does this return 0 for certain txs?
-                'gasPrice': tx_data['gasPrice'],
-                'netFeePaid': tx_data['gasPrice'] * tx_receipt['gasUsed']
-            }
-        
-        transaction_hashes = get_transaction_hashes(block_calls)
-
-        ## Create a new object
-        block = Block(
-            block_number=block_number,
-            data=block_data,
-            receipts=block_receipts_raw,
-            calls=block_calls,
-            logs=block_logs,
-            transaction_hashes=transaction_hashes,
-            txs_gas_data=txs_gas_data,
+        print(
+           f"Cache for block {block_number} did not exist, getting data"
         )
-        
-        ## Write the result to a JSON file for loading in the future
-        write_json(block)
+
+        w3 = Web3(base_provider)
+        block = fetch_block(w3, base_provider, block_number)
+
+        cache_block(cache_path, block)
 
         return block
 
 
+def fetch_block(w3, base_provider, block_number: int) -> Block:
+    ## Get block data
+    block_data = w3.eth.get_block(block_number, True)
+    
+    ## Get the block receipts
+    ## TODO: evaluate whether or not this is sufficient or if gas used needs to be converted to a proper big number.
+    ## In inspect-ts it needed to be converted
+    block_receipts_raw = base_provider.make_request("eth_getBlockReceipts", [block_number])
+
+    ## Trace the whole block, return those calls
+    block_calls_json = w3.parity.trace_block(block_number)
+    block_calls = [
+        BlockCall(**call_json)
+        for call_json in block_calls_json
+    ]
+    
+    ## Get the logs
+    block_hash = (block_data.hash).hex()
+    block_logs = w3.eth.get_logs({'blockHash': block_hash})
+
+    ## Get gas used by individual txs and store them too
+    txs_gas_data = {}
+
+    for transaction in block_data['transactions']:
+        tx_hash = (transaction.hash).hex()
+        tx_data = w3.eth.get_transaction(tx_hash)
+        tx_receipt = w3.eth.wait_for_transaction_receipt(tx_hash)
+        txs_gas_data[tx_hash] = {
+            'gasUsed': tx_receipt['gasUsed'], # fix: why does this return 0 for certain txs?
```

Comment:
`# fix: why does this return 0 for certain txs?`
I think this was Taarush's comment, for visibility for everyone else Erigon has a bug and is returning 0 for the gas used of some transactions :(